apps/system/js/sync_manager.js

```diff
@@ -0,0 +1,547 @@
+ *    'sync': requests a sync on demand. Although we have scheduled
+ *            synchronizations, we allow on demand sync requests.
+ *    'getInfo': gets the current information about the state of Sync
+ *               including the state (enabled, disabled, etc.), the
+ *               last time a successful sync happened and the last error
+ *               if there's any.
+ *
+ * It also triggers one message through the same IAC API with name
+ * 'onsyncchange' when something (state, error, lastSync) changes on Sync.
+ *
+ * SyncManager also communicates with the Sync app via IAC. Every time a
+ * synchronization is required (either because the requestSync timer fires or
+ * because the user requested a sync on demand), we wake the Sync app up and
+ * give it the required information to perform the synchronization that
+ * includes the URL of the Syncto server, a valid FxA assertion and its
+ * associated encryption keys and the list of collections to be synchronized.
+ */
+
+(function() {
+  const FXA_EVENT = 'mozFxAccountsUnsolChromeEvent';
+
+  const SYNC_TASK = 'gaia::system::firefoxsync';
+  const SYNC_REQUEST_IAC_KEYWORD = 'gaia::sync::request';
+  const SYNC_MANAGEMENT_API_IAC_KEYWORD = 'gaia::sync::management';
+
+  const COLLECTIONS = ['history', 'passwords'];
+
+  // Keys of asyncStorage persisted data.
+  const SYNC_STATE = 'sync.state';
+  const SYNC_STATE_ERROR = 'sync.state.error';
+  const SYNC_LAST_TIME = 'sync.lastTime';
+
+  var SyncManager = function() {};
+
+  SyncManager.STATES = [];
+
+  SyncManager.SERVICES = [];
+
+  SyncManager.SETTINGS = [
+    /** READ ONLY SETTINGS **/
+
+    // The user can choose the collections she wants to sync. These settings
+    // indicates the user choice per collection.
+    'sync.collection.history.enabled',
+    'sync.collection.passwords.enabled',
+
+    'sync.server.url',
+    'sync.scheduler.interval',
+    'sync.scheduler.wifionly'
+  ];
+
+  SyncManager.EVENTS = [
+    /** IAC based sync management API **/
+    'iac-' + SYNC_MANAGEMENT_API_IAC_KEYWORD,
+
+    /** Sync State Machine events **/
+    'onsyncdisabled',
+    'onsyncenabled',
+    'onsyncenabling',
+    'onsyncerrored',
+    'onsyncsyncing',
+
+    /** Sync app lifecycle **/
+    'killapp',
+  ];
+
+  SyncManager.SUB_MODULES = [
+    'SyncStateMachine'
+  ];
+
+  BaseModule.create(SyncManager, {
+    name: 'SyncManager',
+    DEBUG: true,
+    EVENT_PREFIX: 'sync',
+    settings: new Map(),
+    store: new Map(),
+
+    /** BaseModule life cycle **/
+
+    _start: function() {
+      return new Promise((resolve, reject) => {
+        LazyLoader.load('/shared/js/sync/errors.js', () => {
+          navigator.mozSetMessageHandler('request-sync', event => {
+            this.debug('Sync request');
+            if (event.task != SYNC_TASK) {
+              return;
+            }
+            Service.request('SyncStateMachine:sync');
+          });
+
+          SyncManager.SETTINGS.forEach(setting => {
```

Comment:
you should not have do to this... https://github.com/mozilla-b2g/gaia/blob/master/apps/system/js/base_module.js#L433-L435